### ISSUE TITLE
Loving the Integration

### ISSUE THREAD

**Issue by rstrouse** [Crewski/njsPC-HA#17]
Hey I am one of the contributors to njsPC.  I am looking to add functionality for the chemistry controllers.  I have noticed that your unique id generation algorithm for status sensors is not always unique.  This is because some of these in njsPC need to be prefixed by the equipment type.

For instance, the status entity for a chlorinator with id = 1 will have the same unique id as a pump with id = 1.  I fixed this in the code by changing the unique_id property for the StatusSensor class in sensor.py to the following.  It includes the event filter that is passed in when the entity is created.  This appears to be `chlorinator`, `pump`, `chemController`, `filter`, ...etc.

```
   @property
    def unique_id(self) -> str:
        """ID of the sensor"""
        return self.coordinator.api.get_unique_id(
            f'status_{self._event}_{self._equipment["id"]}'
        )
```

Still playing around with HA but I have also noticed that the base entity class is not being called to ensure the entity gets hooked up to the device.  The only time this is an issue is when a circuit or feature is removed from njsPC.  In this instance it will throw an exception (keyerror) and not connect the entities.  This is because the event key may not be in the data and checking this is not necessarily required.

It took me a whole lot of head scratching to figure out exactly what this DeviceInfo is for when I created the integration for my Somfy shades, but it simply uses this to match up the data for the entities.  So if the touple for the indentifiers match, then the entity belongs to the device.  This should be returned for every entity not just the base entity.

Do you want some pull requests?

**Comment by Crewski** [comment]
@rstrouse yes, definitely put in any pull requests.  The deviceinfo/base entity stuff was my first attempt at it, and as you noticed isn't implemented correctly.  I am slowly rewriting things to hopefully the proper way (I feel like the custom integration documentation for HA is severely lacking in way of examples), but I'd happily use what you have done.  Thanks.

**Comment by rstrouse** [comment]
The examples they do have really do nothing of any value and there a million HA apis that need to be called in sequence to get a proper integration spun up that actually processes data.  None of them seem to be documented anywhere as to their purpose or when they have callbacks which is even more problematic due to the loopy and thready nature of snake code.  Then again I am just learning Parseltounge.

The other part that makes this tough is that there are wobbly definitions for hub, domain, platform, component, device, entity, and service.  So it ends up not that it isn't very good, its that it is misleading.  For me that took a bit of chewing to figure out.  How is a platform an entity type?

I'll work on getting entities (and platforms) built for tanks, probes, and chem pumps.  Maybe toss schedules in there as well.

**Comment by rstrouse** [comment]
How do you feel about making multiple devices for the controller?  For instance a pump is a device and it has entities.  The same would be true for each piece of equipment.  Each entity can then have its own attributes.

For instance there would potentially be a pool device and a spa device.  Each could have a climate entity as well as a switch entity to turn it on and off.   For something like an IntelliBrite there would be a circuit device that has a theme entity.

For a chlorinator there could be a status entity, setpoint entity, and all the sensor entities that are related to it (salt level, current output ... etc).

**Comment by Crewski** [comment]
I went back and forth on that initially.  I'm definitely not opposed to having each piece of equipment be its own device....and it probably should be that way anyway.  Honestly I don't think the device association is very important.  It was added in because it was requested.  I have all my important information displayed on the dashboard, I would think others probably do the same.

**Comment by rstrouse** [comment]
Ok so I reworked the plugin so it lines up with about 24 devices on my system.  It really makes it cleaner for the integration and the entity naming.  I still have to create the body devices but after that it should be all converted over and the unique ids should be in place.  Basically, it amounted to adding the Chem Controller stuff and re-arranging some of your terrific code to stamp the device info on each entity.  Here are a couple of examples

![image](https://user-images.githubusercontent.com/47839015/216460107-76acc059-cc76-4c36-a0ce-dfdbda2cedaa.png)

![image](https://user-images.githubusercontent.com/47839015/216460276-5d6a61e9-6b6d-4513-ba12-f99bc917eb6a.png)

![image](https://user-images.githubusercontent.com/47839015/216460402-f9b530f9-b4a9-46de-8842-cd6e8bb18c88.png)

![image](https://user-images.githubusercontent.com/47839015/216460518-96cf531b-a496-4122-9781-c4984b725a9b.png)

**Comment by Crewski** [comment]
Wow, that's looking great.  I definitely see how having multiple devices is the way to go.  My pool controller is very simples (Pump, Heater, Lights) so there weren't too many entities....but they definitely add up.

Only thing I'm noticing is a typo in the software version for the device where there isn't an opening parenthesis.

**Comment by rstrouse** [comment]
I have a pull request to you with my changes.

I converted it all over to device centric functionality and added all the stuff for chem controllers, filters, bodies, and binary sensors.  I do want to add discovery for ssdp and mdns in the mix.  I have the code but until I get around to publishing via HACS locally I cannot test the config flow.

**Comment by Crewski** [comment]
Great, I'll look it over.  Discovery would be fantastic!  You should be able to just dump all your code into the custom-components folder to bypass HACS and reboot.

**Comment by rstrouse** [comment]
I did it in my dev core so that option is a bit hobbled in there when you create it as a non custom integration.  I'll add the config_flow this weekend since I can bang it from my fork.  If I changed anything you do not like it was unintended.  I tried to keep all the entities the same with the exception of changes to the new HA units formatting and making sensible ids.  The rest of the changes are equipment additions and new sensors/controls.
